### PR TITLE
feat: [Validation] add `required_with_value` validation rules

### PR DESCRIPTION
**Pull request by warcooft**
**Description**
Add `required_with_value` validation rules.

Example:
```php
$data = [
    'is_internal'     => '1',
    'identity_number' => ''
];

$rules = [
    'identity_number' => 'required_with_value[is_internal,1]'
];

dd($this->validateData($data, $rules)); // returns false
 
/**
 * Result:
 * 'identity_number' => 'The identity_number field is required when field and value has is_internal,1 is present.'
 */
``` 

If this rule are worth adding to the framework, I will proceed to write the documentation.

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by kenjis** [comment]
Thank you for sending this PR!

We expect the following in all Pull Requests (PRs).
- PRs must be sent to the [appropriate branch](https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#branching)
- All git commits must be [GPG-signed](https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#signing)
- Must follow our [style guide](https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#php-style)
- Be [commented](https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#comments) in the PHP source file
- Be documented in the [user guide](https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#user-guide)
- Be [unit tested](https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#unit-testing)
- Pass all checks in GitHub Actions

> [!IMPORTANT]
> We expect all code changes or bug-fixes to be accompanied by one or more tests added to our test suite to prove the code works.

If pull requests do not comply with the above, they will likely be closed. Since we are a team of volunteers, we don't have any more time to work on the framework than you do. Please make it as painless for your contributions to be included as possible.

See https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md

**Comment by kenjis** [comment]
@warcooft 
Can you show us use cases for this rule?

**Comment by warcooft** [comment]
`required_with_value`, is a rule that typically checks if a field is required based on the presence of another field and a specific value in that field.

In this case, the rule `'identity_number' => 'required_with_value[is_internal,1]'` means that the `identity_number` field is required if the `is_internal` field exists and its value is “1”.

Use Case Explanation:
1. Internal User Identification:
    * Suppose you have a form where users can register, and there's an option to indicate whether they are internal users (is_internal flag).
    * If `is_internal` is set to "1", it indicates that the user is internal.
    * In such a case, you want to ensure that an `identity_number` is provided because internal users might have an associated identification number that is required for validation or access purposes.
2. Conditional Validation:
    * This validation rule ensures that the `identity_number` field is only validated (required) when `is_internal` is explicitly set to "1". If `is_internal` is not present or is set to a different value (like '0' or '2'), the `identity_number` field is not required.

## Here are a few more use cases where the `required_with_value` validation rule can be applied effectively:

Use Case 1: Optional Contact Information
```php
$data = [
    'is_business'     => '1',
    'contact_email'   => '',
    'contact_phone'   => ''
];

$rules = [
    'contact_email' => 'required_with_value[is_business,1]',
    'contact_phone' => 'required_with_value[is_business,1]'
];

// Validation result depends on 'is_business' flag
dd($this->validateData($data, $rules));

``` 
- Scenario: A form for business entities where users can indicate if they are registering as a business (is_business).
- Validation Rule: `contact_email` and `contact_phone` are required only if `is_business` is set to 1. This ensures that businesses provide at least one method of contact.

Use Case 2: Additional Document Upload
```php
$data = [
    'is_verified'      => '1',
    'verification_doc' => ''
];

$rules = [
    'verification_doc' => 'required_with_value[is_verified,1]'
];

// Validation ensures document is uploaded if 'is_verified' is true
dd($this->validateData($data, $rules));

``` 
- Scenario: A verification process where users can indicate if they want their account verified (is_verified).
- Validation Rule: `verification_doc` (a document upload field) is required if `is_verified` is set to 1. This ensures that verified accounts have necessary documentation submitted.

Use Case 3: Special Permissions
```php
$data = [
    'user_type'     => 'admin',
    'special_access' => ''
];

$rules = [
    'special_access' => 'required_with_value[user_type,admin]'
];

// Admin users may need to specify special access rights
dd($this->validateData($data, $rules));
``` 
- Scenario: A user management system where administrators (user_type) can have additional permissions or access (special_access).
- Validation Rule: `special_access` is required if `user_type` is set to 'admin'. This ensures that administrators define their special access settings if applicable.

Use Case 4: Membership Levels
```php
$data = [
    'is_premium'   => '1',
    'membership_id' => ''
];

$rules = [
    'membership_id' => 'required_with_value[is_premium,1]'
];

// Premium members must provide their membership ID
dd($this->validateData($data, $rules));
``` 
- Scenario: A membership platform where users can upgrade to premium (is_premium).
- Validation Rule: `membership_id` is required if `is_premium` is set to 1. This ensures that premium members provide their membership ID for verification and access to premium features.

Use Case 5: Subscription Services
```php
$data = [
    'subscribe_newsletter' => '1',
    'email_preferences'    => ''
];

$rules = [
    'email_preferences' => 'required_with_value[subscribe_newsletter,1]'
];

// Users who subscribe to newsletter must set email preferences
dd($this->validateData($data, $rules));
``` 
- Scenario: A newsletter subscription form where users opt-in to receive newsletters (subscribe_newsletter).
- Validation Rule: `email_preferences` is required if `subscribe_newsletter` is set to 1. This ensures that users who subscribe to the newsletter also provide their email preferences (e.g., frequency, content preferences).

Summary:
The `required_with_value` rule is versatile and useful in scenarios where certain fields depend on the presence or value of other fields. It ensures that our application's forms collect necessary information conditionally based on user selections or data states, enhancing data integrity and user experience by guiding users to provide relevant information when needed.

I didn't think about creating a separate package because the resources are too small so it would be quite effective if this rule could be adopted. Thank you

**Comment by warcooft** [comment]
This rule is inspired from this discussion post:
https://github.com/codeigniter4/shield/discussions/1137#discussioncomment-9880685
https://github.com/codeigniter4/shield/discussions/1137#discussioncomment-9880478

in that case, the appropriate rule  is `required_with_value` instead of `permit_empty`.
```php
'eid_number' => [
            'label' => 'EID No',
            'rules' => [
                // 'permit_empty',
                'required_with_value[is_internal_employee,1]',
                'max_length[18]',
                'is_not_unique[employees_master.eid_no]',
            ],
            'errors' => [
                'is_not_unique' => 'EID Number [ {value} ] doest not exist.',
            ],
        ]
``` 

it is clearer to guide users to provide relevant information. without using that rules internal users or employees could miss it because they accidentally left the column blank. 

I also applied this rule to my registration form for user type filtering.

**Comment by ddevsr** [comment]
This rules same behaviour on `in_list[value1]`, no reason to add this

**Comment by kenjis** [comment]
> This rules same behaviour on `in_list[value1]`

No. `in_list` fails if field is not within a predetermined list.
https://codeigniter4.github.io/CodeIgniter4/libraries/validation.html#rules-for-general-use

**Comment by warcooft** [comment]
> This rules same behaviour on `in_list[value1]`, no reason to add this

Validation rules `required_with_value` and `in_list` serve different purposes in the context of data validation.
 
This rule ensures that a field is required if **another field has a certain value**. while this `in_list` rule checks whether the value of a field is included in the specified list of values.

**Comment by warcooft** [comment]
Hi @kenjis, @ddevsr,  
Can you confirm for me whether this rule is worth including in the framework? If feasible, I will try move this PR to 4.6 branch then work on documentation and testing.

**Comment by ddevsr** [comment]
I know what you mean, I have a habit of validating if a certain type needs additional information like below. And i can check multiple criteria for additional information.

```php
// ......
$params_type = $this->request->getPost('params_type');

$validationRules = [
            'params_name'       => 'required|string|max_length[150]',
            'params_email'      => 'required|valid_email|max_length[100]',
            'params_phone'      => 'required|numeric|max_length[15]',
            'params_fax'        => 'required|numeric|max_length[15]',
            'params_website'    => 'required|string|max_length[100]',
            'params_lob'        => 'required|integer',
            'params_type'       => 'required|string|in_list[BUSINESS,PERSONAL]',
            'params_region'     => 'required|string|max_length[13]',
            'params_address'    => 'required|string|max_length[253]',
            'params_pic'        => 'required',
            'params_created'    => 'required|integer',
];

if($params_type === 'BUSINESS'){
            $validationRules['params_identity'] = 'required|numeric|max_length[16]',
}

$this->validation->setRules($validationRules, $validationMessages);
```

**Comment by kenjis** [comment]
@ddevsr Thanks. I got what you mean.

**Comment by ddevsr** [comment]
@kenjis @warcooft I think no need ya, sending POST which is not needed can increase process time

**Comment by kenjis** [comment]
@warcooft  How do we validate the identity_number format?

```php
        $data = [
            'is_internal'     => '0',
            'identity_number' => '',
        ];

        $rules = [
            'is_internal'     => 'in_list[0,1]',
            'identity_number' => 'required_with_value[is_internal,1]|integer'
        ];

        d($this->validateData($data, $rules)); // false
        dd($this->validator->getErrors());     // "The identity_number field must contain an integer."
```

**Comment by warcooft** [comment]
> I know what you mean, I have a habit of validating if a certain type needs additional information like below. And i can check multiple criteria for additional information.
> 
> ```
> // ......
> $params_type = $this->request->getPost('params_type');
> 
> $validationRules = [
>             'params_name'       => 'required|string|max_length[150]',
>             'params_email'      => 'required|valid_email|max_length[100]',
>             'params_phone'      => 'required|numeric|max_length[15]',
>             'params_fax'        => 'required|numeric|max_length[15]',
>             'params_website'    => 'required|string|max_length[100]',
>             'params_lob'        => 'required|integer',
>             'params_type'       => 'required|string|in_list[BUSINESS,PERSONAL]',
>             'params_region'     => 'required|string|max_length[13]',
>             'params_address'    => 'required|string|max_length[253]',
>             'params_pic'        => 'required',
>             'params_created'    => 'required|integer',
> ];
> 
> if($params_type === 'BUSINESS'){
>             $validationRules['params_identity'] = 'required|numeric|max_length[16]',
> }
> 
> $this->validation->setRules($validationRules, $validationMessages);
> ```
> 
> If we write rules `required_with_value` is `must` validate and we must send POST field `params_identity` no matter what a fill field `params_type`

@ddevsr I'm the opposite, I prefer to make it a rule, it's clearer and easier to remember or gives me a clue when there is a similar case.

**Comment by warcooft** [comment]
> I know what you mean, I have a habit of validating if a certain type needs additional information like below. And i can check multiple criteria for additional information.
> 
> ```
> // ......
> $params_type = $this->request->getPost('params_type');
> 
> $validationRules = [
>             'params_name'       => 'required|string|max_length[150]',
>             'params_email'      => 'required|valid_email|max_length[100]',
>             'params_phone'      => 'required|numeric|max_length[15]',
>             'params_fax'        => 'required|numeric|max_length[15]',
>             'params_website'    => 'required|string|max_length[100]',
>             'params_lob'        => 'required|integer',
>             'params_type'       => 'required|string|in_list[BUSINESS,PERSONAL]',
>             'params_region'     => 'required|string|max_length[13]',
>             'params_address'    => 'required|string|max_length[253]',
>             'params_pic'        => 'required',
>             'params_created'    => 'required|integer',
> ];
> 
> if($params_type === 'BUSINESS'){
>             $validationRules['params_identity'] = 'required|numeric|max_length[16]',
> }
> 
> $this->validation->setRules($validationRules, $validationMessages);
> ```
> 
> If we write rules `required_with_value` is `must` validate and we must send POST field `params_identity` no matter what a fill field `params_type`

again, yes this is quite simple when done on a controller but it is a bit tricky the way this is implemented in some cases.
E.g. using conditioning in shield validation if the source is from config\Validation.php, conditioning validation if the source is from the models.

**Comment by michalsn** [comment]
IMO this is a good idea to include. I have a rule with the same name in one of my projects.

The only change I would like to see is the addition of support for multiple parameters. Something like this should be possible:
```php
$rules = [
    'special_option' => 'required_with_value[normal_option,1,2]'
];
```
This will make it even more flexible.

**Comment by warcooft** [comment]
> @warcooft How do we validate the identity_number format?
> 
> ```
>         $data = [
>             'is_internal'     => '0',
>             'identity_number' => '',
>         ];
> 
>         $rules = [
>             'is_internal'     => 'in_list[0,1]',
>             'identity_number' => 'required_with_value[is_internal,1]|integer'
>         ];
> 
>         d($this->validateData($data, $rules)); // false
>         dd($this->validator->getErrors());     // "The identity_number field must contain an integer."
> ```

@kenjis  Of course, to solve this problem we can combine it with `permit_empty` rule. but we need to modify the `processPermitEmpty` method in `system/Validation/Validation.php line:440` and include `require_with_value` for this to work properly. How about you?

**Comment by warcooft** [comment]
> IMO this is a good idea to include. I have a rule with the same name in one of my projects.
> 
> The only change I would like to see is the addition of support for multiple parameters. Something like this should be possible:
> 
> ```
> $rules = [
>     'special_option' => 'required_with_value[normal_option,1,2]'
> ];
> ```
> 
> This will make it even more flexible.

Great idea, i agree for that.

**Comment by warcooft** [comment]
Hi guys, which syntax is more appropriate to use below?

```php
// style #1
$rules = [
            'normal_option'  => 'in_list[0,1,2]',
            'special_option' => 'required_with_value[normal_option,1,2]|integer|permit_empty'
        ];

// style #2
$rules = [
            'normal_option'  => 'in_list[0,1,2]',
            'special_option' => 'required_with_value[normal_option:1,2]|integer|permit_empty' 
        ];
```

**Comment by michalsn** [comment]
@warcooft I would stay with the comma syntax, as it is used in other rules.

**Comment by christianberkman** [comment]
I actually also have a use case for the opposite, let's call it `only_with` where I only permit (but not require) a field like `date_left` if `has_left` has a value. I believe I solved that in the controller though.

**Comment by warcooft** [comment]
@michalsn If I prefer the second one, it is clearer to differentiate between fields and values. but, I would also like to hear other opinions.

**Comment by warcooft** [comment]
@michalsn I have an idea. 

Should we refactor the name of this rule to better describe what this function actually does? 

For example, we could use `required_if` instead of `required_with_value`. 

What do you think about this?

**Comment by michalsn** [comment]
@warcooft Personally, I'm okay with changing the rule name.

Let's wait to see if others also find this rule useful enough.

@christianberkman `allow_with` should not be a big problem to implement - it will be a normal rule (if I'm not mistaken). The `required_*` rules are a bit different because they require special treatment in the process of validation.

**Comment by christianberkman** [comment]
> @christianberkman `allow_with` should not be a big problem to implement - it will be a normal rule (if I'm not mistaken). The `required_*` rules are a bit different because they require special treatment in the process of validation.

Depending if the the referenced field is checked for any value or a specific value. In my case, `has_left` could be `null`, `0`, or `1` so I would need to check the value. But may be it better solved in the controller. For me I delete the `date_left` field from the data array if `has_left != 1` before I validate the data array.

**Comment by warcooft** [comment]
@christianberkman Actually there are some rules that can also be completed in the controller. The benefits provided by implementing it in a rule are that our code will be neater, easier to read, and easier to maintain. I also don't think this is a big problem to implement.

**Comment by kenjis** [comment]
Thank you for updating.

I don't oppose to add this rule because 1) it seems there are many use cases, and 2) it is a special rule that needs to change `Validation` class.

Laravel has `required_if` and it seems the same behavior. So I'm okay with `required_if`.
https://laravel.com/docs/11.x/validation#rule-required-if

File: system/Language/en/Validation.php [hunk 1]
```diff
@@ -53,6 +53,7 @@
     'required'              => 'The {field} field is required.',
     'required_with'         => 'The {field} field is required when {param} is present.',
     'required_without'      => 'The {field} field is required when {param} is not present.',
+    'required_with_value'   => 'The {field} field is required when field and value has {param} is present.',
```
Comment (kenjis) [inline system/Language/en/Validation.php:56]: This message is too bad. Please change to one that is easy to understand for humans.

> The identity_number field is required when field and value has is_internal,1 is present.

Comment (warcooft) [reply system/Language/en/Validation.php:56]: how about this one
`'required_if' => 'The {field} field is required when the field specified in {param} is present.',`
> The identity_number field is required when the field specified in is_internal,1 is present. 

or 
`'required_if' => 'The {field} field is required when the {param} field has the specified value.',`
> The identity_number field is required when the is_internal,1 field has the specified value.

Comment (warcooft) [reply system/Language/en/Validation.php:56]: or this one:

'required_if' => 'The {field} field is required when the {param} field has the given value.',
> The identity_number field is required when the is_internal,1 field has the given value.

Comment (kenjis) [reply system/Language/en/Validation.php:56]: I think this is good.

> The identity_number field is required when the is_internal field is the given value.

**Comment by kenjis** [comment]
First, please fix the base branch of this PR branch.
https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/workflow.md#if-you-sent-to-the-wrong-branch

**Pull request closed without merging**